_rules/auto-update-text-efbfc7.md

```diff
@@ -0,0 +1,190 @@
+---
+id: efbfc7
+name: Text content that updates automatically can be paused, stopped or hidden
+rule_type: atomic
+description: |
+  This rule checks that there are mechanisms to pause, stop or hide the auto-updating of text content.
+accessibility_requirements: # Remove whatever is not applicable
+  wcag20:2.2.2: # Pause, Stop, Hide (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
+  - DOM Tree
+  - CSS Styling
+acknowledgements:
+  authors:
+    - Carlos Duarte
+---
+
+## Applicability
+
+The rule applies to any [visible][] [HTML element][] in an [HTML document][] if:
+
+- the `innerText` property of the [element][html element] changes without the user [activating][activate] or [focusing][focus] any [element][html element] in the same [HTML document][];
```

Comment:
I'm still having an uneasy feeling about the causality, but I'm not entirely sure how to express it… (sorry looks like this is going to be a long rant…)

---
As far as I understand, part of the problem that 2.2.2 detects is that the auto-updates never end. It is not a one time event but a continuous thing that can be distracting. So, something like this
```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<body onload="setTimeout(change, 1000)">
  <p>Random number: <span id="target">1</span></p>
</body>
```
should be good because there is only 1 update. There is no need to have a mechanism to stop the auto-updates because they stop themselves. This is currently applicable (content changes without user activating or focusing anything) and I'd say it fails the rule (there is no UI component to stop the change).

The first point that raises is that there is no real difference between an auto-update every second and a single update after 1, 2, 3, …, 3600 seconds… (`onload="for(var i=1; i<=3600; i++) { setTimeout(change, i*1000) }"`) There is something on the SC which is about the global perception (all these updates are part of the same auto-update, not individual events). That does not bode well for making an objective applicability…

The "auto-update" part of the SC does not have any time limit (contrarily to the "blinking" part which is OK if less than 5 seconds). But maybe we need to enforce it to make the applicability objective ("the time between the first and last (if any) change of the content is 5 seconds or more")?

---
Then, there is the causality between the click (or other user action) and the changes far in the future. If I look at Inapplicable example 2, I have no doubt that this is OK for WCAG. But the click is going to trigger updates far in the future. The 3600th update, happening 1 hour from now, may well look like it happens without the user activating an element. That is going to be OK because we are going to perceive all the updates as a single event and thus keep the causality in mind.

But if the updates are considered individually, it won't be that obvious that they happen due to that click. That hints that the "user activating" part is not really objective and unambiguous.

```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<p>Random number: <span id="target">1</span></p>
<input type="button" id="control" onclick="setTimeout(startUpdates, 5*60*1000)" value="Click me if you dare!" />
```
I think that is OK for WCAG (the updates start due to user interaction). It should also be inapplicable for the rle (the updates start on user activating an element). But I am not sure that the perception is going to be that the updates are linked to the click (because of the 5 minutes delay before it). And if I feel that the updates are not linked to the click, then this is applicable and fails the rule (but passes WCAG).

Compare with:
```html
<script type="text/javascript" src="/test-assets/efbfc7/script.js"></script>
<body onload="setTimeout(startUpdates, 5*60*1000)">
  <p>Random number: <span id="target">1</span></p>
  <input type="button" id="control" value="Click me if you dare!" />
</body>
```
this second case is bad for WCAG and for the rule.
But if I don't have access to the code (or if code is obfuscated or the triggers hidden behind undecidable conditions), the difference between the two is very slim. Sure, I can wait in the first case and make enough experiences to convince me that the updates start as a consequence of me clicking the button. But that is not obvious and might be dangerously close to subjectivity (which is not allowed in Applicability).

---
What I'm trying to figure out here is how to make both the "action => updates" causality, and the "all these updates are a single thingie" unambiguous and objective.

Reply:
I guess I'm not reading the SC the same way you do:

- The SC does not apply only to auto-updates that never end. One example of an auto-updating content in the Understanding document is an "auto-advancing presentation". I certainly expect a presentation to end sometime :-)

- My interpretation of the SC is that it aims at making sure that users have the time they need to perceive and understand the content before it automatically updates (that is why a solution that changes the frequency of the updates also passes the SC - so we should take that into account, I guess) @